pkg/kubectl/cmd/util/cmdtools.go

```diff
@@ -0,0 +1,265 @@
+/*
+Copyright 2015 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"math/rand"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"strings"
+
+	"github.com/golang/glog"
+
+	"k8s.io/kubernetes/pkg/kubectl/util/term"
+)
+
+const (
+	// TODO: on Windows rely on 'start' to launch the editor associated
+	// with the given file type. If we can't because of the need of
+	// blocking, use a script with 'ftype' and 'assoc' to detect it.
+	defaultEditor = "vi"
+	defaultShell  = "/bin/bash"
+	windowsEditor = "notepad"
+	windowsShell  = "cmd"
+	gitDiff       = "git diff --no-index"
+	normalDiff    = "diff"
+)
+
+type CmdName string
+
+const (
+	EditorCmd CmdName = "editor"
+	DiffCmd   CmdName = "diff"
+)
+
+// CmdTools include
+// editor and diff tools
+type CmdTool struct {
+	Name  CmdName
+	Args  []string
+	Shell bool
+}
+
+// NewDefaultCmdTool creates a struct CmdTool that uses the OS environment to
+// locate the editor or diff program, looking at environment variable to find
+// the proper command line. for editor: If the provided editor has no spaces,
+// or no quotes, it is treated as a bare command to be loaded.
+// Otherwise, the string will be passed to the user's shell for execution.
+func NewDefaultCmdTool(name CmdName, envs []string) CmdTool {
+	exec, shell := defaultEnv(name, envs)
+	return CmdTool{
+		Name:  name,
+		Args:  exec,
+		Shell: shell,
+	}
+}
+
+func defaultEnvShell() []string {
+	shell := os.Getenv("SHELL")
+	if len(shell) == 0 {
+		shell = platformize(defaultShell, windowsShell)
+	}
+	flag := "-c"
+	if shell == windowsShell {
+		flag = "/C"
+	}
+	return []string{shell, flag}
+}
+
+func defaultEnv(cmdName CmdName, envs []string) ([]string, bool) {
+	var cmd string
+	for _, env := range envs {
+		if len(env) > 0 {
+			cmd = os.Getenv(env)
+		}
+		if len(cmd) > 0 {
+			break
+		}
+	}
+	if len(cmd) == 0 {
+		switch cmdName {
+		case EditorCmd:
+			cmd = platformize(defaultEditor, windowsEditor)
+		case DiffCmd:
+			cmd = checkGitDiffInstalled(gitDiff, normalDiff)
+		}
+	}
+
+	if !strings.Contains(cmd, " ") {
+		return []string{cmd}, false
+	}
+	if !strings.ContainsAny(cmd, `"'\`) {
+		return strings.Split(cmd, " "), false
+	}
```

Comment:
There are more than three ways to create a string literal in bash. This can be trivially broken.

For instance:

```bash
MY_ENV=<<-EOF
this is a string with space characters in it
EOF
```

In general, I do not like using `strings` functions on shell-related literals. There are too many edge cases.
Is there a better way to do this?